entity-service/src/main/kotlin/com/egm/stellio/entity/service/EntityService.kt

```diff
@@ -717,21 +717,30 @@ class EntityService(
         neo4jRepository.deleteEntity(entityId)
 
     @Transactional
-    fun deleteEntityAttribute(entityId: String, attributeName: String, contextLink: String): Boolean {
+    fun deleteEntityAttribute(entityId: String, attributeName: String, datasetId: URI?, deleteAll: Boolean, contextLink: String): Boolean {
         val expandedAttributeName = expandJsonLdKey(attributeName, contextLink)!!
 
-        if (neo4jRepository.hasPropertyOfName(EntitySubjectNode(entityId), expandedAttributeName))
-            return neo4jRepository.deleteEntityProperty(entityId, expandedAttributeName) >= 1
-        else if (neo4jRepository.hasRelationshipOfType(
-            EntitySubjectNode(entityId),
-            expandedAttributeName.toRelationshipTypeName())
-        )
-            return neo4jRepository.deleteEntityRelationship(
-                entityId,
-                expandedAttributeName.toRelationshipTypeName()
-            ) >= 1
-
-        throw ResourceNotFoundException("Attribute $attributeName not found in entity $entityId")
+        if (deleteAll) {
+            if (neo4jRepository.hasPropertyOfName(EntitySubjectNode(entityId), expandedAttributeName))
+                return neo4jRepository.deleteEntityProperty(
+                    subjectNodeInfo = EntitySubjectNode(entityId), propertyName = expandedAttributeName, deleteAll = deleteAll) >= 1
+            else if (neo4jRepository.hasRelationshipOfType(
+                    EntitySubjectNode(entityId), expandedAttributeName.toRelationshipTypeName()))
+                return neo4jRepository.deleteEntityRelationship(subjectNodeInfo = EntitySubjectNode(entityId),
+                    relationshipType = expandedAttributeName.toRelationshipTypeName(), deleteAll = deleteAll) >= 1
+            throw ResourceNotFoundException("Attribute $attributeName not found in entity $entityId")
+        } else {
+            if (neo4jRepository.hasPropertyInstance(EntitySubjectNode(entityId), expandedAttributeName, datasetId))
+                return neo4jRepository.deleteEntityProperty(EntitySubjectNode(entityId), expandedAttributeName, datasetId, deleteAll) >= 1
+            else if (neo4jRepository.hasRelationshipInstance(
+                    EntitySubjectNode(entityId), expandedAttributeName.toRelationshipTypeName(), datasetId))
+                return neo4jRepository.deleteEntityRelationship(
+                    EntitySubjectNode(entityId), expandedAttributeName.toRelationshipTypeName(), datasetId, deleteAll) >= 1
+            datasetId?.let { throw
+            ResourceNotFoundException("Instance with datasetId $datasetId of attribute $attributeName not found in entity $entityId") }
+
+            throw ResourceNotFoundException("Default instance of $attributeName not found in entity $entityId")

```

Comment:
Quite a strange construct. A basic `if (datasetId != null) throw ... else throw ...` would be more readable.